crates/full-node/sov-db/src/historical_state.rs

```diff
@@ -1,70 +1,134 @@
 use std::fmt::Debug;
+use std::sync::Arc;
 
 use rockbound::cache::delta_reader::DeltaReader;
+use rockbound::versioned_db::{VersionedDeltaReader, VersionedSchemaBatch};
 use rockbound::{SchemaBatch, SchemaKey, SchemaValue};
 use sov_rollup_interface::common::SlotNumber;
 
 use crate::metrics::StateMaterializationMetrics;
 use crate::namespaces::{KernelNamespace, Namespace, UserNamespace};
-use crate::schema::namespace::StateValues;
+use crate::schema::namespace::NomtStateValues;
 use crate::schema::tables::StateRootHashes;
 use crate::schema::types::StateRootHashId;
-use crate::{ensure_version_is_correct, DbOptions};
+use crate::DbOptions;
 
 const STATE_ROOT_HASH_SINGLETON: StateRootHashId = StateRootHashId(0);
 
 /// A typed wrapper around the [`DeltaReader`] for reading materializing historical rollup state.
 #[derive(Debug, Clone)]
 pub struct HistoricalStateReader {
     /// The underlying [`DeltaReader`] correctly routes requests to previous snapshots and/or [`rockbound::DB`]
-    db: DeltaReader,
+    user: VersionedDeltaReader<NomtStateValues<UserNamespace>>,
+    /// The underlying [`DeltaReader`] correctly routes requests to previous snapshots and/or [`rockbound::DB`]
+    kernel: VersionedDeltaReader<NomtStateValues<KernelNamespace>>,
+    other: DeltaReader,
+
     /// The [`SlotNumber`] that will be used for the next batch of writes to the DB
     /// This [`SlotNumber`] is also used for querying data,
     /// so if this instance of [`HistoricalStateReader`] is used as read-only, it won't see newer data.
     next_version: SlotNumber,
 }
 
+/// A collection of changes to the state db. Includes versioned changes to user/kernel state, and a plain schema batch of changes to any other columns.
+#[derive(Debug, Clone, Default)]
+pub struct StateChanges {
+    pub(crate) user: Arc<VersionedSchemaBatch<NomtStateValues<UserNamespace>>>,
+    pub(crate) kernel: Arc<VersionedSchemaBatch<NomtStateValues<KernelNamespace>>>,
+    pub(crate) other: Arc<SchemaBatch>,
+}
+
 impl HistoricalStateReader {
     const DB_PATH_SUFFIX: &'static str = "historical_state";
     const DB_NAME: &'static str = "historical-state-db";
 
-    /// Create a new instance of [`HistoricalStateReader`] from a given [`DeltaReader`].
-    pub fn with_delta_reader(reader: DeltaReader) -> anyhow::Result<Self> {
-        let next_version = Self::next_version_from(&reader)?;
-        tracing::trace!(?next_version, "Initialized historical state reader");
-        Ok(Self {
-            db: reader,
+    // Used for testing only.
+    #[allow(dead_code)]
```

Comment:
If it is only for testing, let's make it `cfg(test)` or `cfg(feature = "test-utils")